### PR TITLE
constant variables

### PR DESCRIPTION
**Pull request by LianaHus**
added isConstant to VariableDeclaration

### CODE REVIEW

File: libsolidity/AST.h [hunk 1]
```diff
@@ -465,15 +474,17 @@ class VariableDeclaration: public Declaration
...
 protected:
 	Visibility getDefaultVisibility() const override { return Visibility::Internal; }
 
 private:
 	ASTPointer<TypeName> m_typeName;    ///< can be empty ("var")
-	ASTPointer<Expression> m_value;		///< the assigned value, can be missing
+	ASTPointer<Expression> m_value;     ///< the assigned value, can be missing
 	bool m_isStateVariable;             ///< Whether or not this is a contract state variable
 	bool m_isIndexed;                   ///< Whether this is an indexed variable (used by events).
+	bool m_isConstant;                  ///< Whether the variable is const
```
Comment (chriseth) [inline libsolidity/AST.h]: "const" -> "is a compile-time constant"?

File: libsolidity/AST.h [hunk 2]
```diff
@@ -440,13 +440,22 @@ class FunctionDefinition: public Declaration, public VariableScope, public Docum
...
+	VariableDeclaration(
+		SourceLocation const& _location,
+		ASTPointer<TypeName> const& _type,
+		ASTPointer<ASTString> const& _name,
+		ASTPointer<Expression> _value,
+		Visibility _visibility,
+		bool _isStateVar = false,
+		bool _isIndexed = false,
+		bool _isConstant = false):
 			Declaration(_location, _name, _visibility),
```
Comment (chriseth) [inline libsolidity/AST.h]: I thik this block should only be indented one tab after `VariableDeclaration` now, while I would indent the block above one further tab not to align with this block here.

File: libsolidity/AST.cpp [hunk 3]
```diff
@@ -197,7 +197,7 @@ vector<pair<FixedHash<4>, FunctionTypePointer>> const& ContractDefinition::getIn
 				}
 
 			for (ASTPointer<VariableDeclaration> const& v: contract->getStateVariables())
-				if (v->isPublic() && functionsSeen.count(v->getName()) == 0)
+				if (v->isPublic() && functionsSeen.count(v->getName()) == 0 && !v->isConstant())
```
Comment (chriseth) [inline libsolidity/AST.cpp]: Please add a virtual function "isPartOfExternalInterface" to `Declaration`.

File: libsolidity/AST.cpp [hunk 4]
```diff
@@ -323,7 +323,7 @@ string FunctionDefinition::getCanonicalSignature() const
 bool VariableDeclaration::isLValue() const
 {
 	// External function parameters are Read-Only
```
Comment (chriseth) [inline libsolidity/AST.cpp]: Please update comment.

File: libsolidity/AST.h [hunk 5]
```diff
@@ -465,15 +474,17 @@ class VariableDeclaration: public Declaration
...
 	bool isIndexed() const { return m_isIndexed; }
+	bool isConstant() const { return m_isConstant; }
 
 protected:
 	Visibility getDefaultVisibility() const override { return Visibility::Internal; }
 
 private:
 	ASTPointer<TypeName> m_typeName;    ///< can be empty ("var")
-	ASTPointer<Expression> m_value;		///< the assigned value, can be missing
+	ASTPointer<Expression> m_value;     ///< the assigned value, can be missing
```
Comment (chriseth) [inline libsolidity/AST.h]: You changed spaces to a tab here.

File: libsolidity/ExpressionCompiler.cpp [hunk 6]
```diff
@@ -843,8 +843,11 @@ void ExpressionCompiler::endVisit(Identifier const& _identifier)
 	}
 	else if (FunctionDefinition const* functionDef = dynamic_cast<FunctionDefinition const*>(declaration))
 		m_context << m_context.getVirtualFunctionEntryLabel(*functionDef).pushTag();
-	else if (dynamic_cast<VariableDeclaration const*>(declaration))
-		setLValueFromDeclaration(*declaration, _identifier);
+	else if (VariableDeclaration const* variable = dynamic_cast<VariableDeclaration const*>(declaration))
```
Comment (chriseth) [inline libsolidity/ExpressionCompiler.cpp]: You can use `auto` here.

File: libsolidity/ExpressionCompiler.cpp [hunk 7]
```diff
@@ -843,8 +843,11 @@ void ExpressionCompiler::endVisit(Identifier const& _identifier)
...
+	{
+		if (!variable->isConstant())
```
Comment (chriseth) [inline libsolidity/ExpressionCompiler.cpp]: And else I think you should visit the value of the variable.

File: libsolidity/Parser.cpp [hunk 8]
```diff
@@ -152,13 +153,21 @@ ASTPointer<ContractDefinition> Parser::parseContractDefinition()
 			VarDeclParserOptions options;
 			options.isStateVariable = true;
 			options.allowInitialValue = true;
+			options.isDeclaredConst = isDeclaredConst;
```
Comment (chriseth) [inline libsolidity/Parser.cpp]: `allowConstant`?

File: test/SolidityEndToEndTest.cpp [hunk 9]
```diff
@@ -3181,6 +3181,27 @@ BOOST_AUTO_TEST_CASE(pass_dynamic_arguments_to_the_base_base_with_gap)
...
+	compileAndRun(sourceCode);
+	BOOST_CHECK(callContractFunction("getX()") == encodeArgs(56));
+}
+
+BOOST_AUTO_TEST_CASE(const_state_variable)
+{
+	char const* sourceCode = R"(
+		contract Foo {
+			function getX() constant returns (uint r) { return x; }
+			constant uint x = 56;
```
Comment (chriseth) [inline test/SolidityEndToEndTest.cpp]: The syntax should be `uint constant x = 56` (just like `uint public x = 56`).

File: libsolidity/Parser.cpp [hunk 10]
```diff
@@ -152,13 +153,21 @@ ASTPointer<ContractDefinition> Parser::parseContractDefinition()
...
 			stateVariables.push_back(parseVariableDeclaration(options));
+			isDeclaredConst = false;
 			expectToken(Token::Semicolon);
 		}
 		else if (currentToken == Token::Modifier)
 			modifiers.push_back(parseModifierDefinition());
 		else if (currentToken == Token::Event)
 			events.push_back(parseEventDefinition());
+		else if (currentToken == Token::Const)
```
Comment (chriseth) [inline libsolidity/Parser.cpp]: Parsing the constant token should go inside parseVariableDeclaration (see below).

File: test/SolidityEndToEndTest.cpp (hunk 9 above)
Comment (LianaHus) [inline test/SolidityEndToEndTest.cpp]: ah, I took it from the description of the story at pivotal tracker. I'll change it

**Comment by LianaHus** [comment]
rebase

File: libsolidity/AST.h [hunk 11]
```diff
@@ -415,6 +416,7 @@ class FunctionDefinition: public Declaration, public VariableScope, public Docum
 			getVisibility() >= Visibility::Internal;
 	}
 	virtual TypePointer getType(ContractDefinition const*) const override;
+	virtual bool isPartOfExternalInterface() const override { return isPublic() && m_isConstructor && getName().empty(); }
```
Comment (chriseth) [inline libsolidity/AST.h]: Should be `isPublic() && !m_isConstructor && !getName().empty();` (two negations)

File: libsolidity/ExpressionCompiler.cpp [hunk 12]
```diff
@@ -843,8 +843,11 @@ void ExpressionCompiler::endVisit(Identifier const& _identifier)
 	}
 	else if (FunctionDefinition const* functionDef = dynamic_cast<FunctionDefinition const*>(declaration))
 		m_context << m_context.getVirtualFunctionEntryLabel(*functionDef).pushTag();
-	else if (dynamic_cast<VariableDeclaration const*>(declaration))
-		setLValueFromDeclaration(*declaration, _identifier);
+	else if (auto variable = dynamic_cast<VariableDeclaration const*>(declaration))
+	{
+		if (!variable->isConstant())
```
Comment (chriseth) [inline libsolidity/ExpressionCompiler.cpp]: `else variable->getValue()->accept(*this);`

File: libsolidity/AST.h (hunk 11 above)
Comment (LianaHus) [inline libsolidity/AST.h]: thanks I already find that :) will fix

File: libsolidity/ExpressionCompiler.cpp (hunk 12 above)
Comment (LianaHus) [inline libsolidity/ExpressionCompiler.cpp]: sure I don't forget it, I just didn't push last changes

**Comment by chriseth** [comment]
As discussed:
1. Please disallow the constant modifier during parsing for everything except for state variables.
2. During type checking, verify that a constant state variable always has an initializing value and the type of the variable is one of Integer, Enum or FixedString.

**Comment by LianaHus** [comment]
rebase

File: libsolidity/AST.h [hunk 13]
```diff
@@ -440,13 +442,22 @@ class FunctionDefinition: public Declaration, public VariableScope, public Docum
...
 public:
-	VariableDeclaration(SourceLocation const& _location, ASTPointer<TypeName> const& _type,
-							ASTPointer<ASTString> const& _name, ASTPointer<Expression> _value,
-							Visibility _visibility,
-							bool _isStateVar = false, bool _isIndexed = false):
-			Declaration(_location, _name, _visibility),
-			m_typeName(_type), m_value(_value),
-			m_isStateVariable(_isStateVar), m_isIndexed(_isIndexed) {}
+	VariableDeclaration(
+			SourceLocation const& _location,
```
Comment (gavofyork) [inline libsolidity/AST.h]: only a single tab.

Comment (chriseth) [inline libsolidity/AST.h]: We need a clear distinction between parameters and initializer list. Therefore I would say either single tab and `)` unindented on the next line or the way it is here.

Comment (LianaHus) [inline libsolidity/AST.h]: I would say the initialization list should not be unindented. If you remember in the previews ci It was 1 tab for argument list and 2 for member initialization list but I agree that this way it looks better.
Should I change it or leave it like this guys?

**Comment by LianaHus** [comment]
rebased

File: libsolidity/Parser.cpp [hunk 14]
```diff
@@ -327,7 +328,16 @@ ASTPointer<VariableDeclaration> Parser::parseVariableDeclaration(
 		isIndexed = true;
 		m_scanner->next();
 	}
+	if (token == Token::Const)
+	{
+		solAssert(_options.isStateVariable, "");
+		if (m_scanner->peekNextToken() != Token::Identifier && !Token::isElementaryTypeName(m_scanner->peekNextToken()))
```
Comment (chriseth) [inline libsolidity/Parser.cpp]: What is the purpose of this check?

**Pull request merged**